Update Doorbell Siren 6 to support Sound/Volume control

Aeotec-ccheng opened the pull request:
- Added setting/preference option to configure volume/sound via Multichannel Cmd Encap (Sound Switch Configuration SET)

Aeotec-ccheng committed:
Update aeotec-doorbell-siren-6.groovy

- Added setting/preference option to configure volume/sound via Multichannel Cmd Encap (Sound Switch Configuration SET)

cbaumler commented on devicetypes/smartthings/aeotec-doorbell-siren-6.src/aeotec-doorbell-siren-6.groovy:
```diff
@@ -66,10 +110,24 @@ def installed() {
+def soundControl(Endpoint, Volume, Sound) {
+	if (Endpoint && Volume && Sound) {
+    		log.debug "soundControl($Endpoint, $Volume, $Sound)"
+        	encap(zwave.multiChannelV3.multiChannelCmdEncap(sourceEndPoint:0, destinationEndPoint:Endpoint, commandClass:121, command:5, parameter: [Volume,Sound]))
```
I generally see command class values in hexadecimal.

Aeotec-ccheng replied:
@cbaumler 

commandClass = 0x79 //Sound Switch Command Class
command = 0x05 //Configuration SET

greens replied:
@Aeotec-ccheng he's just saying that our typical coding style is to write these params as hex in the code

greens replied:
Also, isn't the intent of the `encap` method to handle this? I don't think you should be passing a multichannel encapsulated command to `encap`

Aeotec-ccheng replied:
@greens - shouldn't be a problem to change these values to hex.

Aeotec-ccheng replied:
i can change the method of how the multichannel cmd encap is handled, but it does need to be secure encapsulated.

I've modified private encap() to properly pass multichannel encapsulation, but if needed, I can move it to a different def/func.

Aeotec-ccheng replied:
@cbaumler @greens - Changed from decimal to hex value use and tested.

Separated Encap function to handle MC cmd encap(sound switch configuration set) separately:
private encap()
private mcEncap()

greens reviewed (changes requested):
A couple of general coding style fixes:

* zwave command classes should consistently be written as hex values
* device preferences and variables should not have initial letters capitalized
* open braces should go on the same line as if statements, and closing braces should be on the same line as following `else`s

Aeotec-ccheng commented:
> A couple of general coding style fixes:
> 
> * zwave command classes should consistently be written as hex values
> * device preferences and variables should not have initial letters capitalized
> * open braces should go on the same line as if statements, and closing braces should be on the same line as following `else`s

Thanks, I'll work on this shortly to follow these coding styles.

Aeotec-ccheng committed:
Update code format

Format fixes
- zwave command classes use hex values for their input
- device preferences and def variables changed to have initial lowercase letters
- open braces are now on same line as if statements
- closing braces are now on the same line as following else statements
- fixed minor formatting issues

Code change
- separated multichannel encap as mcEncap() and encap() to handle Multichannel Cmd Encap (Sound Switch Configuration Set) separately

Aeotec-ccheng commented:
@greens - Finished formatting based on code style, please let me know if any other changes are needed here.

greens commented on devicetypes/smartthings/aeotec-doorbell-siren-6.src/aeotec-doorbell-siren-6.groovy:
```diff
@@ -268,3 +314,29 @@ def keepChildrenOnline() {
+private mcEncap(cmd) {
+	if (cmd) {
+		if (zwaveInfo.zw.contains("s")) {
+			def rawZwaveData = zwave.securityV1.securityMessageEncapsulation().encapsulate(cmd).format()
```
okay, I think I know what you were running into. The context of the function is what was giving you trouble. Based on where you're calling this from, you should just be able to wrap this object in `response()` and return that

Aeotec-ccheng replied:
I'll give this a quick test and change this in the next hour.

Aeotec-ccheng replied:
@greens - Thanks that worked like a charm.

greens commented on devicetypes/smartthings/aeotec-doorbell-siren-6.src/aeotec-doorbell-siren-6.groovy:
```diff
@@ -98,15 +155,17 @@ def zwaveEvent(physicalgraph.zwave.commands.multichannelv3.MultiChannelCmdEncap
 	}
-	def encapsulatedCommand = cmd.encapsulatedCommand()
+    
+	def encapsulatedCommand = cmd.encapsulatedCommand([0x60: 3, 0x79: 1])
```
Can you remove the `0x79: 1` you're passing in?

We haven't updated our backend to support parsing of Sound Switch yet (and there's no handler established in this DTH anyway) so it's not actually doing anything.

Aeotec-ccheng replied:
@greens - Got it, I've removed it from the encapsulatedCommand and secureEncapsulatedCommand.

Aeotec-ccheng committed:
Quick update

Remove 0x79: 1 from secureencapsulation and encapsulation.

Aeotec-ccheng committed:
Condense private mcEncap()

Condensed code to use response(cmd) for MC Cmd Encap (Sound Switch Config SET).

Aeotec-ccheng committed:
just a comment edit

Non-essential commit, changed commend on line 336

Aeotec-ccheng committed:
added return to line 336

- Add return in front of line 336

Aeotec-ccheng committed:
Final update for minor changes 

- Changed steps in description on line 57 for better user flow.

Aeotec-ccheng commented:
I noticed the other endpoints aren't triggering which is caused by state.lastTriggeredSound not changing when controlled via button.

Edited the Multichannel command class to change state.lastTriggeredSound only if NOTIFICATION CC is controlled.

Aeotec-ccheng committed:
Fixed child device status update

- Fixed issue where child devices were not updating. Button press and tamper child devices now change states.
- Edited Line 162 -164 to set state.lastTriggeredSound when a new sound is playing from a different endpoint.

greens commented:
@MGoralczykS want to take a look here?

MGoralczykS commented:
@greens This fixing issue ICP-12798. This looks good.
